operator/buffer/circular_file.go

```diff
@@ -0,0 +1,327 @@
+	f          fileLike
+	seekedRead bool
+	seekedEnd  bool
+	closed     bool
+	readPtrEnd bool
+}
+
+var _ io.ReadWriteCloser = (*circularFile)(nil)
+
+func openCircularFile(filePath string, sync bool, size int64) (*circularFile, error) {
+	fileFlags := os.O_CREATE | os.O_RDWR
+	if sync {
+		fileFlags |= os.O_SYNC
+	}
+
+	f, err := os.OpenFile(filePath, fileFlags, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that the file, if it already existed, is actually the correct size.
+	fsize, err := f.Seek(0, io.SeekEnd)
+	if err != nil {
+		fCloseErr := f.Close()
+		return nil, multierr.Combine(
+			err,
+			fCloseErr,
+		)
+	}
+
+	if fsize == 0 {
+		err := f.Truncate(size)
+		if err != nil {
+			fCloseErr := f.Close()
+			return nil, multierr.Combine(
+				err,
+				fCloseErr,
+			)
+		}
+	} else if fsize != size {
+		fCloseErr := f.Close()
+		return nil,
+			multierr.Combine(
+				fmt.Errorf("configured size (%d) does not match current on-disk size (%d)", size, fsize),
+				fCloseErr,
+			)
+	}
+
+	return &circularFile{
+		Size: size,
+		f:    f,
+	}, nil
+}
+
+func (rb *circularFile) Close() error {
+	if rb.closed {
+		return nil
+	}
+
+	rb.closed = true
+
+	err := rb.f.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read reads from the ring buffer into p, up to len(p) bytes.
+// The contents read are not discarded from the buffer; An independent read pointer
+// is maintained, which can be reset to the start of the buffer using ResetReadOffset()
+func (rb *circularFile) Read(p []byte) (int, error) {
+	if rb.closed {
+		return 0, ErrBufferClosed
+	}
+
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	err := rb.seekReadStart()
+	if err != nil {
+		return 0, err
+	}
+
+	var totalBytesToRead int64
+	var isEOF bool
+	if rb.ReadBytesLeft() < int64(len(p)) {
+		totalBytesToRead = rb.ReadBytesLeft()
+		isEOF = true
+	} else {
+		totalBytesToRead = int64(len(p))
+		isEOF = false
+	}
+
+	var firstReadBytes int64
+	if totalBytesToRead+rb.ReadPtr <= rb.Size {
+		firstReadBytes = totalBytesToRead
+	} else {
+		firstReadBytes = rb.Size - rb.ReadPtr
+	}
+
+	n, err := rb.f.Read(p[:firstReadBytes])
+	rb.ReadPtr = (rb.ReadPtr + int64(n)) % rb.Size
+	if rb.ReadPtr == 0 {
+		rb.seekedRead = false
+	}
+
+	if err != nil {
+		return n, err
+	}
+
+	if firstReadBytes != totalBytesToRead {
+		err = rb.seekReadStart()
+		if err != nil {
+			return n, err
+		}
+
+		n2, err := rb.f.Read(p[firstReadBytes:])
+		rb.ReadPtr += int64(n2)
+		if err != nil {
+			return n + n2, err
+		}
+	}
+
+	if rb.ReadPtr == rb.End && n != 0 {
+		rb.readPtrEnd = true
+	}
+
+	if isEOF {
+		return int(totalBytesToRead), io.EOF
+	}
+
+	return int(totalBytesToRead), nil
+}
+
+// Write writes to the CircularFile.
+// If the end of the file has been reached and no more bytes may be written,
+// io.EOF is returned as an error, as well as the number of bytes that were successfully
+// written.
+func (rb *circularFile) Write(p []byte) (int, error) {
+	if rb.closed {
```

Comment:
Same race condition concern for closing as read. Also, what happens if I call read and write at the same time?